### ISSUE TITLE
WIP: Implementation experience while realizing new API in node-wot

### ISSUE THREAD

**Issue by danielpeintner** [w3c/wot-scripting-api#180]
Note: This is a evolving issue list that gets updated/modified/edited while realizing the new Scripting API in node-wot:

Issues:
* [x] A *light* ThingBuilder functionality like we used to have in `ExposedThing` is desirable
  * e.g., `addProperty("propName", {...}, "initValue")` allowed for initializing a property with an initial value before starting it. With the new API there is no such possibility. I think some basic `addProperty()`, `addAction()`, `addEvent()` methods on ExposedThing would be of of great help in this regard.
    * can be done in helper class
    * initializing of properties is a separate step though 
  * make *clearer* in spec/note that initializing values can be done before exposing a thing by adding `writeProperty()` in ExposedThing example right after `WoT.produce(...)`
* [x] readProperty() parameters/options, see https://github.com/w3c/wot-scripting-api/issues/182
* [x] use of getTD()/getThingDescription() method instead of td attribute (allows overloading, keep internal structure as TD, logging, ... more flexibility), see https://github.com/w3c/wot-scripting-api/issues/184
* [x] handler IDs for unobserving property/event, see https://github.com/w3c/wot-scripting-api/issues/183

**Comment by zolkis** [comment]
Currently one can build the init dictionary that is used for creating an ExposedThing. The API for that is in JavaScript, no need to duplicate an [Object API](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object), IMHO: one can already add (TD) definitions for WoT Property, Action and Event and then they can be used from the created ExposedThing. 

There is no need to "add later", since all properties can be added when creating an ExposedThing, and if we speak about taking a ConsumedThing and extend it, we can do that with providing `produce()` an object that is extended with the desired interaction definitions.

Building objects in any language is provided by the language implementation. The WoT specialization is what should be added by the WoT Scripting API, i.e. how to create an ExposedThing and a ConsumedThing, which should be already covered.

**Comment by danielpeintner** [comment]
As mentioned, with the current approach there is no possibility "for initializing a property with an initial value before starting" the thing.

This is heavily used in all our examples/scenarios.

**Comment by zolkis** [comment]
Init value passing can be resolved in easier ways than building a full object API, if that is an important use case. It's not even clear that would be the case, and we do have the experience it's a big and complex API. 

We could add one more parameter to `expose()`, or have an explicit `init()` method, or add another parameter to the constructor for init values (since the current init dict contains only the TD definitions).

**Comment by zolkis** [comment]
In principle, we need the following when adding interactions to ExposedThing:
- a name (of a property, action or event)
- TD definition
- initial value.

Currently we can add only the TD definitions, with one single call. So a dev needs to build a JS object for the TD, then pass it either to the produce() method, or to the ExposedThing constructor.
There is no possibility to change that later for a given ExposedThing object - which I don't see a big limitation, since "later" is same as "now", all needs to be in the same script and ExposedThing is not active before expose() is called.

Now what we are missing are the init values. As said before, we could pass these init values in a second argument to produce() and the ExposedThing constructor, or the expose() method, but again in "bulk mode", not one by one. (NB: we have a bug in the spec: ExposedThing constructor is missing from Web IDL, but present in prose). 

If we want to break this up into separate addition/removal/editing of individual interactions, then we need to define a full object API, and the gain would be disputable (multiple method calls instead of one). I would like to avoid this option, since we have seen how complex it goes.

However, we could have a simple version of ThingBuilder, like
```javascript
partial interface ExposedThing {
  void addProperty(DOMString name, ThingDescription td, optional any value);
  void removeProperty(name);
  void addAction(DOMString name, ThingDescription td);  // action is set separately
  void removeAction(name);
  void addEvent(DOMString name, ThingDescription td);
  void removeEvent(name);
};
```
But then the individual `ThingDescription` parameters should represent a single Property, or Action, or Event. Also, when is it allowed to add/remove interactions (not while being exposed). Then, how to stop a Thing, modify and re-expose. There is a lot of text to write and make sure the provided `td` parameters are valid, etc etc.

Instead, now we can do the same thing by adding elements to a JS object and then call produce() on it when done. And no lengthy extra spec is needed for equivalent functionality.

**Comment by danielpeintner** [comment]
I am not so much concerned about ThingBuilder functionality (and I would stick to what we have so far). I was just pointing out some experiences/differences/issues I ran into while implementing.

Having said that, I miss
* setting a property value when instantiating an `ExposedThing`
* see many benefits of having `getTD()` method compared to a `td` attribute
* options `readProperty()`

I think those points are best discussed in one of our next scripting calls

**Comment by zolkis** [comment]
@danielpeintner  can you check if #185 and #186 handled this sufficiently, and if yes, please close the issue.

**Comment by danielpeintner** [comment]
@zolkis I checked all checkboxes (hence, all issues resolved so far).

However, I would like to keep this *general* issue open (for the next couple of days) for the case we encounter more. I hope this OK for you?

**Comment by zolkis** [comment]
OK, if you want to keep track of this round's all issues in one issue - just edit the first comment to add a new checkbox and make a related new issue.

**Comment by danielpeintner** [comment]
Note: Some `getTD()` calls left

**Comment by zolkis** [comment]
`getTD()` as well fixed in #187
